### ISSUE TITLE
Ingress with sticky session returns disabled backend servers in the persistence cookie

### ISSUE THREAD

**Issue by zerikv** [haproxytech/kubernetes-ingress#258]
Hello,

The versions installed are : 
- kubectl version : 1.18.3
- Ingress controller : haproxytech/kubernetes-ingress:1.4.10;  installed by Helm with the following options :
```
helm install haproxy-ingress haproxytech/kubernetes-ingress --create-namespace --namespace haproxy-controller --set controller.kind=DaemonSet
```

The sticky session has been enabled in the ingress resource with the folowing annotation :
```
  annotations:
    haproxy.org/cookie-persistence: "INGRESS"	
```

In the `haproxy.cfg` of one Ingress Controller Pod : the cookie parameter has been added in the backend section :
```
	backend prod-service-8080
	  mode http
	  balance roundrobin
	  option forwardfor
	  cookie INGRESS indirect nocache insert
	  server SRV_PjVIU 127.0.0.1:8080 disabled check weight 128 cookie SRV_PjVIU
	  server SRV_4K9eq 127.0.0.1:8080 disabled check weight 128 cookie SRV_4K9eq
	  server SRV_j0noV 127.0.0.1:8080 disabled check weight 128 cookie SRV_j0noV
	  server SRV_JSHL7 127.0.0.1:8080 disabled check weight 128 cookie SRV_JSHL7
	  server SRV_WKhDY 127.0.0.1:8080 disabled check weight 128 cookie SRV_WKhDY
	  server SRV_p97tQ 127.0.0.1:8080 disabled check weight 128 cookie SRV_p97tQ
	  server SRV_FSPct 127.0.0.1:8080 disabled check weight 128 cookie SRV_FSPct
	  server SRV_FXEFc 127.0.0.1:8080 disabled check weight 128 cookie SRV_FXEFc
	  server SRV_lck1S 127.0.0.1:8080 disabled check weight 128 cookie SRV_lck1S
	  server SRV_LhjRT 127.0.0.1:8080 disabled check weight 128 cookie SRV_LhjRT
	  server SRV_zXX0N 127.0.0.1:8080 disabled check weight 128 cookie SRV_zXX0N
	  server SRV_KxASA 127.0.0.1:8080 disabled check weight 128 cookie SRV_KxASA
	  server SRV_92EZ0 127.0.0.1:8080 disabled check weight 128 cookie SRV_92EZ0
	  server SRV_d02dz 127.0.0.1:8080 disabled check weight 128 cookie SRV_d02dz
	  server SRV_kWvD2 127.0.0.1:8080 disabled check weight 128 cookie SRV_kWvD2
	  server SRV_aOAGe 127.0.0.1:8080 disabled check weight 128 cookie SRV_aOAGe
	  server SRV_gqjn9 127.0.0.1:8080 disabled check weight 128 cookie SRV_gqjn9
	  server SRV_pCcjK 127.0.0.1:8080 disabled check weight 128 cookie SRV_pCcjK
	  server SRV_X3Wji 127.0.0.1:8080 disabled check weight 128 cookie SRV_X3Wji
	  server SRV_MdGIA 127.0.0.1:8080 disabled check weight 128 cookie SRV_MdGIA
	  server SRV_SmHsV 127.0.0.1:8080 disabled check weight 128 cookie SRV_SmHsV
	  server SRV_t7KaB 127.0.0.1:8080 disabled check weight 128 cookie SRV_t7KaB
	  server SRV_tISTc 127.0.0.1:8080 disabled check weight 128 cookie SRV_tISTc
	  server SRV_YUgOK 127.0.0.1:8080 disabled check weight 128 cookie SRV_YUgOK
	  server SRV_NYkNy 127.0.0.1:8080 disabled check weight 128 cookie SRV_NYkNy
	  server SRV_cID0Y 127.0.0.1:8080 disabled check weight 128 cookie SRV_cID0Y
	  server SRV_jNHJc 127.0.0.1:8080 disabled check weight 128 cookie SRV_jNHJc
	  server SRV_rIkvf 127.0.0.1:8080 disabled check weight 128 cookie SRV_rIkvf
	  server SRV_E7oyN 127.0.0.1:8080 disabled check weight 128 cookie SRV_E7oyN
	  server SRV_uDkdg 127.0.0.1:8080 disabled check weight 128 cookie SRV_uDkdg
	  server SRV_mUKtw 10.1.104.228:8080 check weight 128 cookie SRV_mUKtw
	  server SRV_IFOAX 10.1.239.9:8080 check weight 128 cookie SRV_IFOAX
	  server SRV_XjBvz 127.0.0.1:8080 disabled check weight 128 cookie SRV_XjBvz
	  server SRV_DZRvc 127.0.0.1:8080 disabled check weight 128 cookie SRV_DZRvc
	  server SRV_R9jjT 127.0.0.1:8080 disabled check weight 128 cookie SRV_R9jjT
	  server SRV_p6hVC 127.0.0.1:8080 disabled check weight 128 cookie SRV_p6hVC
	  server SRV_TOjlQ 127.0.0.1:8080 disabled check weight 128 cookie SRV_TOjlQ
	  server SRV_uWFzQ 127.0.0.1:8080 disabled check weight 128 cookie SRV_uWFzQ
	  server SRV_NgOOd 127.0.0.1:8080 disabled check weight 128 cookie SRV_NgOOd
	  server SRV_3imfJ 127.0.0.1:8080 disabled check weight 128 cookie SRV_3imfJ
	  server SRV_2j42m 127.0.0.1:8080 disabled check weight 128 cookie SRV_2j42m
	  server SRV_z9qHG 127.0.0.1:8080 disabled check weight 128 cookie SRV_z9qHG
```

There are two servers enabled (`SRV_mUKtw` and `SRV_IFOAX`) to forward requests  on the two Pods of the deployment.

When I send a request toward this Ingress Controller the response contains the INGRESS persistence cookie but the server selected (`SRV_92EZ0`) is disabled.
```
$ curl -v http://rd-lin07:31764/gmaoCS02/ -s -o /dev/null
* Connected to rd-lin07 (172.19.10.157) port 31764 (#0)
> GET /gmaoCS02/ HTTP/1.1
> Host: rd-lin07:31764
> User-Agent: curl/7.57.0
> Accept: */*
>
< HTTP/1.1 302
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< set-cookie: JSESSIONID=FDFE84135A0D27EB349C3A3444CF4E1B.production; Path=/gmaoCS02; HttpOnly
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< pragma: no-cache
< expires: 0
< x-content-type-options: nosniff
< location: http://rd-lin07:31764/gmaoCS02/public/login.jsp
< content-length: 0
< date: Wed, 16 Dec 2020 09:48:56 GMT
< set-cookie: INGRESS=SRV_92EZ0; path=/
<
```

If I send another request with the cookies, a new backend server is selected (this time `SRV_aOAGe`) but it's again a disabled server.
```
$ curl -v http://rd-lin07:31764/gmaoCS02/public/login.jsp -b 'JSESSIONID=FDFE84135A0D27EB349C3A3444CF4E1B.production; INGRESS=SRV_92EZ0;' -s -o /dev/null
* Connected to rd-lin07 (172.19.10.157) port 31764 (#0)
> GET /gmaoCS02/public/login.jsp HTTP/1.1
> Host: rd-lin07:31764
> User-Agent: curl/7.57.0
> Accept: */*
> Cookie: JSESSIONID=FDFE84135A0D27EB349C3A3444CF4E1B.production; INGRESS=SRV_92EZ0;
>
< HTTP/1.1 200
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< pragma: no-cache
< cache-control: no-cache
< cache-control: must-revalidate
< cache-control: no-store
< cache-control: max_age=0
< expires: Mon, 8 Aug 2006 10:00:00 GMT
< content-type: text/html;charset=UTF-8
< content-length: 6701
< date: Wed, 16 Dec 2020 09:49:37 GMT
< set-cookie: INGRESS=SRV_aOAGe; path=/
<
```

The requests are load balanced between the two backend servers because with a real access from a web browser thare are many 401 errors.

Do you see what I have to add in the configuration to overcome this problem?

Thank you for your help.

**Comment by Mo3m3n** [comment]
Hi @zerikv 
Thanks for reporting this, a bug related to cookies was fixed yesterday here https://github.com/haproxytech/kubernetes-ingress/commit/d792481622bd35a42f28cf969c947148a2d1b1b1 and is available in nightly from stable branch (v1.4) `haproxytech/kubernetes-ingress:1.4-dev` can you give it a try and see if it fixes your issue ?

**Comment by zerikv** [comment]
Hi Mo3m3n,
Thank you for you quick reply. I have just tried the v1.4-dev :
```
	NAME: haproxy
	LAST DEPLOYED: Wed Dec 16 12:13:55 2020
	NAMESPACE: haproxy-controller
	STATUS: deployed
	REVISION: 1
	TEST SUITE: None
	NOTES:
	HAProxy Kubernetes Ingress Controller has been successfully installed.

	Controller image deployed is: "haproxytech/kubernetes-ingress:1.4-dev".
	Your controller is of a "DaemonSet" kind. Your controller service is running as a "NodePort" type.
```
With this version the behaviour is different but it seems to be buggy yet because the value of the persistence cookie is now its name: `set-cookie: INGRESS=INGRESS;` :
```
$ curl -v http://rd-lin07:30670/gmaoCS02/ -s -o /dev/null
* Connected to rd-lin07 (172.19.10.157) port 30670 (#0)
> GET /gmaoCS02/ HTTP/1.1
> Host: rd-lin07:30670
> User-Agent: curl/7.57.0
> Accept: */*
>
< HTTP/1.1 302
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< set-cookie: JSESSIONID=EF993597AD30B8ADA72B630CB619EFD9.production; Path=/gmaoCS02; HttpOnly
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< pragma: no-cache
< expires: 0
< x-content-type-options: nosniff
< location: http://rd-lin07:30670/gmaoCS02/public/login.jsp
< content-length: 0
< date: Wed, 16 Dec 2020 11:21:43 GMT
< set-cookie: INGRESS=INGRESS; path=/
```

**Comment by Mo3m3n** [comment]
You're right, sorry I have overlooked the issue yesterday and provided an inadequate fix.
So getting back to `haproxytech/kubernetes-ingress:1.4.10`, It seems you get the wrong server on the sticky cookie only when you scale-up or down, but the number of enabled/disabled server do not change this won't happen. Can you confirm this ?

**Comment by zerikv** [comment]
It's not caused by a changement of the number of replicas in the deployment. The configuration of the backend servers in `haproxy.cfg`  didn't change.

**Comment by Mo3m3n** [comment]
I was not able to reproduce this and I always get an enabled server in the persistence cookie.
Can you [enable](https://github.com/haproxytech/kubernetes-ingress/tree/master/documentation#syslog-server) HAProxy logs (to stdout) ?
And provide traffic logs after a couple of tests.

**Comment by zerikv** [comment]
ok I'm going to collect the logs.

The DaemonSet  has deployed 3 Pods. In these 3 haproxy the cookies values affected at each backend server is different. Is it normal? I thought that the mapping "cookie -> backend" should be the same in all ingress controller, shouldn't it ?

```
rd-lin07$ kubectl exec -it pod/haproxy-kubernetes-ingress-md9d4 -n haproxy-controller -- cat /etc/haproxy/haproxy.cfg | grep '^  server' | grep 'cookie' | grep -v disabled
  server SRV_lck1S 10.1.104.228:8080 check weight 128 cookie SRV_lck1S
  server SRV_kWvD2 10.1.239.9:8080 check weight 128 cookie SRV_kWvD2
rd-lin07$ kubectl exec -it pod/haproxy-kubernetes-ingress-hkfxd -n haproxy-controller -- cat /etc/haproxy/haproxy.cfg | grep '^  server' | grep 'cookie' | grep -v disabled
  server SRV_uDkdg 10.1.239.9:8080 check weight 128 cookie SRV_uDkdg
  server SRV_cID0Y 10.1.104.228:8080 check weight 128 cookie SRV_cID0Y
rd-lin07$ kubectl exec -it pod/haproxy-kubernetes-ingress-sw8dp -n haproxy-controller -- cat /etc/haproxy/haproxy.cfg | grep '^  server' | grep 'cookie' | grep -v disabled
  server SRV_IFOAX 10.1.104.228:8080 check weight 128 cookie SRV_IFOAX
  server SRV_gqjn9 10.1.239.9:8080 check weight 128 cookie SRV_gqjn9
```

**Comment by Mo3m3n** [comment]
well you've just found the answer to this issue :+1: 
I was testing with a Daemonset with a single k8s node, so I was not able to reproduce it.
> I thought that the mapping "cookie -> backend" should be the same in all ingress controller, shouldn't it ?

Right it should be but it is the not the case, because the server names are random strings so each haproxy pod has different server names for the same backend, which explains the issue here.

I think we have to change this and name backend servers in a deterministic way.
I will get back to you with more info

**Comment by zerikv** [comment]
The Chart has deployed the following pods and services :
```
$ kubectl get pod -n haproxy-controller -o wide
NAME                                                          READY   STATUS    RESTARTS   AGE   IP             NODE       NOMINATED NODE   READINESS GATES
haproxy-kubernetes-ingress-b6jfj                              1/1     Running   0          54m   10.1.239.23    rd-lin07   <none>           <none>
haproxy-kubernetes-ingress-default-backend-85c64d7688-66dpm   1/1     Running   0          54m   10.1.239.22    rd-lin07   <none>           <none>
haproxy-kubernetes-ingress-default-backend-85c64d7688-zvbqh   1/1     Running   0          54m   10.1.104.233   rd-lin06   <none>           <none>
haproxy-kubernetes-ingress-mvzdp                              1/1     Running   0          54m   10.1.209.163   rd-lin05   <none>           <none>
haproxy-kubernetes-ingress-q78b6                              1/1     Running   0          54m   10.1.104.234   rd-lin06   <none>           <none>

$ kubectl get svc -n haproxy-controller -o wide
NAME                                         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                     AGE   SELECTOR
haproxy-kubernetes-ingress                   NodePort    10.100.210.115   <none>        80:32419/TCP,443:30029/TCP,1024:31988/TCP   61m   app.kubernetes.io/instance=haproxy,app.kubernetes.io/name=kubernetes-ingress
haproxy-kubernetes-ingress-default-backend   ClusterIP   None             <none>        8080/TCP                                    61m   app.kubernetes.io/instance=haproxy,app.kubernetes.io/name=haproxy-kubernetes-ingress-default-backend
```

You can found attached [issue-258-cfg-logs.zip](https://github.com/haproxytech/kubernetes-ingress/files/5704057/issue-258-cfg-logs.zip) the configuration and logs  of the HAProxy running in the 3 pods  :
- pod-haproxy-kubernetes-ingress-b6jfj.cfg
- pod-haproxy-kubernetes-ingress-mvzdp.cfg
- pod-haproxy-kubernetes-ingress-q78b6.cfg

The test send a first GET toward the ingress controller installed on node rd-lin07 :
```
$ curl -v http://rd-lin07:32419/gmaoCS02/ -s
* Connected to rd-lin07 (172.19.10.157) port 32419 (#0)
> GET /gmaoCS02/ HTTP/1.1
> Host: rd-lin07:32419
> User-Agent: curl/7.57.0
> Accept: */*
>
< HTTP/1.1 302
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< set-cookie: JSESSIONID=5CE973EE19AFCF17396288A6B74D7BDA.production; Path=/gmaoCS02; HttpOnly
< cache-control: no-cache, no-store, max-age=0, must-revalidate
< pragma: no-cache
< expires: 0
< x-content-type-options: nosniff
< location: http://rd-lin07:32419/gmaoCS02/public/login.jsp
< content-length: 0
< date: Wed, 16 Dec 2020 14:40:12 GMT
< set-cookie: INGRESS=SRV_pCcjK; path=/
<
```

The log attached `pod-haproxy-kubernetes-ingress-b6jfj.log` trace this first GET :
```
172.19.10.157:12219 [16/Dec/2020:14:40:12.796] http prod-s1234567-service-8080/SRV_pCcjK 0/0/0/1/1 302 440 - - --NI 1/1/0/0/0 0/0 "GET rd-lin07:32419/gmaoCS02/ HTTP/1.1"
```

btw the client IP field (%ci) seems incorrect because it's the IP address of the host :
```
$ kubectl get node -o wide
NAME       STATUS   ROLES    INTERNAL-IP   
rd-cent7   Ready    master   172.19.10.135 
rd-lin05   Ready    <none>   172.19.10.195 
rd-lin06   Ready    <none>   172.19.12.123 
rd-lin07   Ready    <none>   172.19.10.157 
```
The second GET is send toward the same Ingress Controller : 
```
$ curl -v http://rd-lin07:32419/gmaoCS02/public/login.jsp -s -b 'JSESSIONID=5CE973EE19AFCF17396288A6B74D7BDA.production; INGRESS=SRV_pCcjK;' -o /dev/null
* Connected to rd-lin07 (172.19.10.157) port 32419 (#0)
> GET /gmaoCS02/public/login.jsp HTTP/1.1
> Host: rd-lin07:32419
> User-Agent: curl/7.57.0
> Accept: */*
> Cookie: JSESSIONID=5CE973EE19AFCF17396288A6B74D7BDA.production; INGRESS=SRV_pCcjK;
>
< HTTP/1.1 200
< x-frame-options: SAMEORIGIN
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< pragma: no-cache
< cache-control: no-cache
< cache-control: must-revalidate
< cache-control: no-store
< cache-control: max_age=0
< expires: Mon, 8 Aug 2006 10:00:00 GMT
< content-type: text/html;charset=UTF-8
< content-length: 6701
< date: Wed, 16 Dec 2020 14:40:54 GMT
< set-cookie: INGRESS=SRV_t7KaB; path=/
<
```

But this second GET is logged in another Pod on node  rd-lin05 (see pod-haproxy-kubernetes-ingress-mvzdp.log)  :
```
10.1.239.0:57936 [16/Dec/2020:14:40:54.943] http prod-s1234567-service-8080/SRV_t7KaB 1/0/1/7/9 200 7048 - - --DI 1/1/0/0/0 0/0 "GET rd-lin07:32419/gmaoCS02/public/login.jsp HTTP/1.1"
```
The client IP is a virtual address of node rd-lin07.

We known the backend servers selected are wrong because they are disabled.

But this forward between 2 ingress controllers is it a normal behaviour? Or is it a side effect of the bug about the disabled backend server?

**Comment by Mo3m3n** [comment]
> btw the client IP field (%ci) seems incorrect because it's the IP address of the host

I think this is because "Packets sent to Services with Type=NodePort are source NAT'd by default. " as stated here https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport, so you won't get the original IP in the logs. You can find in the same link solutions to preserve the source IP.

> But this forward between 2 ingress controllers is it a normal behaviour? 

Yes it is, well you want multiple Ingress Controller instances (one pod per node) so it is legit that your request will balanced between multiple ingress controllers right ?

**Comment by zerikv** [comment]
Thank you for the explanation about the NAT.

Yes you are rigth, I realize that the requests are routed throught the service `haproxy-kubernetes-ingress` which as you explain load balance between the Ingress Controller pods. We have two levels of load balancing and even third with the external LB in the DMZ.

**Comment by Mo3m3n** [comment]
So the current solution for this issue is to rely on [backend-config-snippet](https://github.com/haproxytech/kubernetes-ingress/tree/master/documentation#config-snippet)  
As explained in [documentation](https://github.com/haproxytech/kubernetes-ingress/tree/master/documentation#cookie-persistence) there is a limitation for the current `cookie-persistence` annotation and the alternative is to use instead the following config:
```
backend-config-snippet: |
    dynamic-cookie-key ahgh5kiM
    cookie SRV insert indirect nocache dynamic
```
More about the dynamic cookie configuration is available in [haproxy documentation](https://cbonte.github.io/haproxy-dconv/2.3/configuration.html#4.2-cookie) (scroll to dynamic part)

This is currently available via `haproxytech/kubernetes-ingress:v1.5.0-rc2` and I get back to you this week when `v1.5.0` is published.

**Comment by Mo3m3n** [comment]
The solution discussed above can now be used via `haproxytech/kubernetes-ingress:1.5.0`
